File: lib/dpapp/os/dppip.c
#include "dpapp/dpefd.h"
#include <errno.h>
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

typedef struct
{
    char* path;
    int unlink_on_del; // 0 or 1
} dppip_udata_t;

// 以指定模式打开命名管道 (FIFO)
static dpret_t _dppip_vopen(void* udata, va_list varg)
{
    const char* fifo_path = va_arg(varg, const char*);
    int ev = va_arg(varg, int);
    int unlink_on_del = va_arg(varg, int);

    if (!fifo_path) {
        return DPE_INVAL;
    }

    // 确保 FIFO 存在，不存在则创建；路径存在但不是 FIFO 则报错
    struct stat st;
    if (stat(fifo_path, &st) != 0) {
        if (errno == ENOENT) {
            if (mkfifo(fifo_path, 0666) != 0 && errno != EEXIST) {
                return -errno;
            }
        } else {
            return -errno;
        }
    } else {
        if (!S_ISFIFO(st.st_mode)) {
            return DPE_INVAL;
        }
    }

    // 将 DPEVT 映射为打开模式（仅允许 IN 或 OUT 之一）
    int open_mode = 0;
    if (ev == DPEVT_IN) {
        open_mode = O_RDONLY;
    } else if (ev == DPEVT_OUT) {
        open_mode = O_WRONLY;
    } else {
        return DPE_INVAL;
    }

    // 以推导模式打开 FIFO
    int fd = open(fifo_path, open_mode);
    if (fd < 0) {
        return -errno; // 例如 EACCES, ENOENT, ENXIO
    }

    // 保存用户数据以便清理
    dppip_udata_t* pip = (dppip_udata_t*)udata;
    if (pip) {
        pip->path = strdup(fifo_path);
        if (pip->path == NULL) {
            close(fd);
            return -ENOMEM;
        }
        pip->unlink_on_del = unlink_on_del ? 1 : 0;
    }

    return fd;
}

static void _dppip_fini(void* udata)
{
    dppip_udata_t* pip = (dppip_udata_t*)udata;
    if (!pip)
        return;
    if (pip->path) {
        if (pip->unlink_on_del) {
            struct stat st;
            if (lstat(pip->path, &st) == 0 && S_ISFIFO(st.st_mode)) {
                unlink(pip->path);
            }
        }
        free(pip->path);
        pip->path = NULL;
    }
}

static dpret_t _dppip_copy(void* dst_udata, const void* src_udata)
{
    dppip_udata_t* src = (dppip_udata_t*)src_udata;
    dppip_udata_t* dst = (dppip_udata_t*)dst_udata;
    if (!dst) {
        return DPE_INVAL;
    }
    dst->path = NULL;
    dst->unlink_on_del = 0; // 复制时始终忽略 unlink 标志
    if (src && src->path) {
        dst->path = strdup(src->path);
        if (!dst->path) {
            return DPE_NOMEM;
        }
    }
    return DPE_OK;
}

static dpele_type_t _dppip_type = {
    .name = "pip",
    .type = DPELE_TYPE_EFD,
    .size = sizeof(dppip_udata_t),
    .iotype = DPAIO_TYPE_GFD,
    .events = DPEVT_IN | DPEVT_OUT,
    .init = _dppip_vopen,
    .copy = _dppip_copy,
    .fini = _dppip_fini,
};

// 导出类型函数
const dpele_type_t* dppip_type()
{
    return &_dppip_type;
}

const char* dppip_path(dpefd_t* efd)
{
    if (dpele_type(efd) != dppip_type()) {
        return NULL;
    }
    dppip_udata_t* pip = (dppip_udata_t*)dpele_aux_data(efd);
    return pip->path;
}

File: lib/dplua/dppip.lua
-- 命名管道（FIFO），通过 dpefd / dpele_new(dppip_type(), path, evmask, unlink_on_del) 创建
--
local dpret = require("dpret")
local ffi = require("ffi")

ffi.cdef([[
const dpele_type_t* dppip_type();
const char* dppip_path(dpefd_t* efd);
]])

local C = ffi.C
local ffi_cast = ffi.cast

local DPE_OK = dpret.OK
local DPE_INVAL = dpret.INVAL

local M = {}

M.type = C.dppip_type()

--- 打开一个 FIFO 作为可轮询的事件元素。
--- @param fifo_path string 管道路径（必要时通过 mkfifo 创建）
--- @param evmask_ number 可选；默认 dplua.EVT_IN（读）或使用 dplua.EVT_OUT（写）
--- @param unlink_on_del_ 为 true 时，元素删除时 unlink 路径（默认 false）
--- @return ele, err
function M.open(fifo_path, evmask_, unlink_on_del_)
    if not fifo_path or type(fifo_path) ~= 'string' then
        return nil, DPE_INVAL
    end

    local evmask = ffi_cast("int", tonumber(evmask_) or dplua.EVT_IN)
    local unlink_on_del = ffi_cast("int", unlink_on_del_ == true and 1 or 0)
    local ele = C.dpele_new(M.type, fifo_path, evmask, unlink_on_del)
    if ele == nil then
        return nil, ffi.errno()
    end
    return ele, DPE_OK
end

function M.path(efd)
    if not efd then
        return nil
    end
    local p = C.dppip_path(efd)
    if p == nil then
        return nil
    end
    return ffi.string(p)
end

return M
